### PR TITLE
Keywords buffers v1

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
I do not think there is any for having keyword for each logged field (ish)

Describe changes:
- Adds multiple keywords

List of sufferable keywords is found with
`jq '[paths(.type?=="string")] | .[] | join(".")' etc/schema.json`

This is a draft to get feedback :
- there are more keywords to do like `smb.kerberos.realm`
- To do so, just run `setup-simple-detect-rustb.sh smb-realm` and then implement the rust function `rs_smb_tx_get_realm`
- Maybe more thought can be given about wether the keyword is to client or to server (or both) and at what stage of the transaction it can be used
- There is no documentation added... Should there be ?
- Should we put these new files in a `detect` subdirectory ?

After the buffers, there are more keywords to create for integers, and maybe for objects (such as `smb.ntlmssp`)

Thoughts ?

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on ips_afp_drop_chk.
  

Pipeline 7986

File: src/detect-tftp-file.c [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        InspectionBufferSetup(det_ctx, list_id, buffer, data, data_len);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+void DetectTFTPFileRegister(void)
+{
+    sigmatch_table[DETECT_AL_TFTP_FILE].name = "tftp.file";
```
Comment (victorjulien) [inline src/detect-tftp-file.c:62]: what does this offer over `alert tftp ... file.name; content:"..."; ...`?

**Comment by victorjulien** [comment]
Think this looks good overall. It does show how much boilerplate code there still is per keyword, which we should try to reduce in the not too distant future.

I would like to see a ticket per keyword added, as well as SV tests for each of the keywords.

**Comment by catenacyber** [comment]
> how much boilerplate code there still is per keyword, which we should try to reduce in the not too distant future 

I will try to look what I can reduce further.

Do you have more precise ideas already ?

**Comment by victorjulien** [comment]
> Do you have more precise ideas already ?

For Rust based keywords I think we have talked about a derive syntax, where you could annotate fields in the transaction implementations and then that would lead to all the registration happening automagically. It would be a bit involved as we'd have to account for direction and progress values, but in theory it should be possible.

File: scripts/setup-simple-detect-rustb.sh [hunk 2]
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env bash
+#
+# Script to setup a new 'simple' detect module.
+# Written by Victor Julien <victor@inliniac.net>
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply scripts/setup-simple-detect-rustb.sh:4]: as you wish.

I was too lazy to change that line

(or rather waited some feedback to work further on this)

**Comment by catenacyber** [comment]
> > Do you have more precise ideas already ?
> 
> For Rust based keywords I think we have talked about a derive syntax, where you could annotate fields in the transaction implementations and then that would lead to all the registration happening automagically. It would be a bit involved as we'd have to account for direction and progress values, but in theory it should be possible.

@jasonish do you have thoughts on this ?
Will it remove C code from this ?

I can see this well for `TFTPTransaction.filename`

Will it work for `SNMPTransaction.usm` which is an `Option<String>` ?

And can it work for the more complex smb.domain whose getter from transaction is
```
    match tx.type_data {
        Some(SMBTransactionTypeData::SESSIONSETUP(ref x)) => {
            if let Some(ref ntlmssp) = x.ntlmssp {
                *buf = ntlmssp.domain.as_ptr();
```

I have also seen the case in DHCP where the same field from the transaction gets logged with a different key based on another field value

**Comment by catenacyber** [comment]
> I would like to see a ticket per keyword added, as well as SV tests for each of the keywords.

Spin-off in #7590

**Comment by jasonish** [comment]
> @jasonish do you have thoughts on this ?
> Will it remove C code from this ?

I haven't thought about it enough.

Its good to lookup at when we've done it for 3 or 4 protocols and the code is identical except for the types that you might be urged to do a macro.  Then it makes sense to look at deriving the implementation.  The actual type of the macro shouldn't matter too much..

```
struct Foo {
    #[suri_buffer(name = request_id)]
    pub request_id: u32,
    #[suri_buffer(name = usm)]
   pub usm: Some<String>,
}
```

type stuff shouldn't be a problem.  As we're operating on parsed Rust code  we could generate different code for different types, within reason.

But I didn't do any of our outher generic or derive macros until a very common pattern evolved.

I don't see how it would remove C code.

**Comment by jasonish** [comment]
> registration happening automagically

Registration wouldn't happen automatically.  All a derive macro can do is create functions, so we'd still have to call those functions from somewhere.

**Comment by catenacyber** [comment]
Thanks Jason.

I understand that we can optimize more to save the 12 lines of rust code for function `rs_snmp_tx_get_usm`+ one more line in the logging function

But that does not prevent from having the 81 lines C file + doc + ticket + S-V test where the bigger part of the work is...

That is the derive stuff will not automagically add 100 new keywords to Suricata signatures, so I will not dig into this right now...

**Comment by catenacyber** [comment]
Closing this, new smaller PRs come one by one (cf the current one on DHCP)

**Pull request closed without merging**